yt_dlp/extractor/tumblr.py

```diff
@@ -107,116 +237,170 @@ class TumblrIE(InfoExtractor):
+                    'hd video url', default='', group='options'),
+                video_id, fatal=False)
+            if options:
+                duration = int_or_none(options.get('duration'))
+
+                hd_url = options.get('hdUrl')
+                if hd_url:
+                    # there are multiple formats; extract them
+                    # (and ignore other sources of width/height data as they may be wrong)
+                    sources = []
+                    sd_url = self._search_regex(
+                        r'<source[^>]+src=(["\'])(?P<url>.+?)\1', iframe,
+                        'sd video url', default=None, group='url')
+                    if sd_url:
+                        sources.append((sd_url, 'sd'))
+                    sources.append((hd_url, 'hd'))
+
+                    formats = [{
+                        'url': video_url,
+                        'format_id': format_id,
+                        'height': int_or_none(self._search_regex(
+                            r'_(\d+)\.\w+$', video_url, 'height', default=None)),
+                        'quality': quality,
+                    } for quality, (video_url, format_id) in enumerate(sources)]
+
+        if 'url' not in media_json and not video_url and not iframe_url:
+            # external video host (but we weren't able to figure it out from the api)
             iframe_url = self._search_regex(
                 r'src=["\'](https?://safe\.txmblr\.com/svc/embed/inline/[^"\']+)["\']',
                 webpage, 'embed iframe url', default=None)
             return self.url_result(iframe_url or redirect_url, 'Generic')
 
-        iframe = self._download_webpage(
-            iframe_url, video_id, 'Downloading iframe page',
-            headers={'Referer': redirect_url})
-
-        duration = None
-        sources = []
-
-        sd_url = self._search_regex(
-            r'<source[^>]+src=(["\'])(?P<url>.+?)\1', iframe,
-            'sd video url', default=None, group='url')
-        if sd_url:
-            sources.append((sd_url, 'sd'))
-
-        options = self._parse_json(
-            self._search_regex(
-                r'data-crt-options=(["\'])(?P<options>.+?)\1', iframe,
-                'hd video url', default='', group='options'),
-            video_id, fatal=False)
-        if options:
-            duration = int_or_none(options.get('duration'))
-            hd_url = options.get('hdUrl')
-            if hd_url:
-                sources.append((hd_url, 'hd'))
-
-        formats = [{
-            'url': video_url,
-            'ext': 'mp4',
-            'format_id': format_id,
-            'height': int_or_none(self._search_regex(
-                r'/(\d{3,4})$', video_url, 'height', default=None)),
-            'quality': quality,
-        } for quality, (video_url, format_id) in enumerate(sources)]
-
+        formats = formats or [{
+            'url': media_json.get('url', video_url),
```

Comment:
```suggestion
            'url': media_json.get('url') or video_url,
```
`media_json['url']` can be `None`